### PR TITLE
WIP: Support for layered transport state

### PR DESCRIPTION
**Pull request by yadij**
Stage 2 of the layered transaction model initiated with MasterXaction
class creation.

This stage implements the actual layering mechanism for transport
protocol changes on a socket. Initial layers TCP, TLS, and PROXY
are labeled for client initiated sockets. Squid initiated sockets are
not labeled.

### CODE REVIEW

File: src/MasterXaction.h [hunk 1]
```diff
@@ -41,25 +41,45 @@ class MasterXaction : public RefCountable
 public:
     typedef RefCount<MasterXaction> Pointer;
 
-    explicit MasterXaction(const XactionInitiator anInitiator) : initiator(anInitiator) {};
+    explicit MasterXaction(const XactionInitiator anInitiator, const char *name = nullptr) : layerName(name), initiator(anInitiator) {}
+
+    /// Create a new transaction context/sandbox for a nested
+    /// sub-protocol which inherits all state information of
+    /// the current transaction but leaves the original as read-only.
+    Pointer spawnChildLayer(const char *name) const;
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/MasterXaction.h:49]: > > This is the model I have been designing since the very beginning.
> 
> I do not know why you are saying that, but this assertion cannot address my concern.

I am stating it to clarify that conflicts arising form use of "master" word in the object naming are probably a result of miscommunication earlier about the object currently called MasterXaction. IMO, a more correct name for a *single* instance of this object would probably be more generic (eg ProtocolXaction).


> The master transaction concept is real, useful, used, and should stay with us.

No disagreement from me. Though they ways I see the *use(s)* of it at presently do not align well with the class MasterXaction naming and use.


> This PR appears to go against that concept by abusing master transaction for layers.

Only if you equate a single class MasterXaction instance as being everything that can be represented in "master transaction".

To me "master transaction" is the entire tree of state data that can be obtained given only a Pointer to an MXA object within that tree. The spawn operation is *growing* the master state to include a sub-protocol layer beyond the existing one(s). Which is why I have carefully named this method and avoided the word "master" in its documentation - it is purely about layering.


> The concept of a layer is also real and useful (in many areas), but a layer is not a master transaction.
> 

Agreed. "master" concept AFAIK permits the possible existence of many transactions and layers that combine to form a "master transaction".


> In general, I discourage folks from coding serious architectural changes affecting the entire Squid without at least some form of a preliminary agreement.

Agreed. I thought we had consensus so far as this change goes :( 


> 
> > With MXA currently only storing tcpClient details we can ignore it
> 
> Does `MXA` stand for MasterXaction in your paragraphs?

yes.

> Do you propose to ignore MasterXaction? What do you mean by that?

No. I mean the conflict between data members in MXA and ALE can be ignored for this PR. Server classes and code within them should not yet be making use of MXA beyond performing the spawn correctly. I might adjust logformat logic to make use of the layers for display within this scope, but only after we agree on how to implement those layers.

My planned "stage 3" future PR(s) should be moving Server state data/pointers from being ALE member to MXA member and code to use MXA directly where possible, or via ALE->xaction where not possible. Eventually (stage 4+) ALE becomes empty and logging is various way(s) to display MXA content as text.

> 
> FWIW, MasterXaction stores more than tcpClient details today. And until the ALE/MasterXaction conflict is resolved, MasterXaction may even be viewed as representing a lot of information currently stored in ALE.
> 
> If we cannot agree on the architecture or even terminology, there may be another way to address my concern. If you can explain to me what this PR is meant to accomplish from Squid functionality changes, perhaps I can suggest a way to implement those changes without violating (my understanding of) the MasterXaction concept.

Hopefully the above clears up anywhere we may not have aligned correctly so far. 

To summarize the design reasoning:

 * This PR is meant to enable logic in the class Server objects to know which MXA they should (in future) store their state in.

 * Since Server objects are highly involved with layering one "master transaction" must be able to be represented as a collection of layers and/or sequence of transactions.

For discussion clarity:  I am not using the term nor concept of "sub-master". I am using "sub-protocol" and sub-layer in reference to the layering property of protocol transactions _within_ a "master".

[the remark this answers is not in the document]
Comment (rousskov) [reply src/MasterXaction.h:49]: > conflicts arising form use of "master" word in the object naming are probably a result of miscommunication earlier about the object currently called MasterXaction. IMO, a more correct name for a *single* instance of this object would probably be more generic (eg ProtocolXaction).

IMO, MasterXaction name reflects the intended purpose of that class. We disagree here. IIRC, my opinion has not changed since that class was added, so it is possible that this (previously hidden) disagreement goes back to 2013.


>> Alex: This PR appears to go against that concept by abusing master transaction for layers.

> Amos: Only if you equate a single class MasterXaction instance as being everything that can be represented in "master transaction".

A class and an instance of a class share the same primary meaning or concept. An instance is just a specific embodiment or example of that concept. Let's not attempt to separate the two when we are discussing design-level decisions. Such a separation feels completely meaningless to me.


> To me "master transaction" is the entire tree of state data that can be obtained given only a Pointer to an MXA object within that tree. 

For a meaningful/usable definition, one has to color generic words like "state" and "data" with some specificity: State of what? What kind of data? I also find it very difficult to understand a definition of "master transaction" that is based on a MasterXaction class that itself has a disputed meaning. As the result, I cannot tell whether my definition of a master transaction concept and the MasterXaction class matches yours, but, in hope to make progress, I can assume that it does.

> The spawn operation is *growing* the master state to include a sub-protocol layer beyond the existing one(s). Which is why I have carefully named this method and avoided the word "master" in its documentation - it is purely about layering.

* A master transaction may include or contain many individual protocol transactions and possibly other layers. "Spawn" would _not_ be the best verb for naming creation of those parts and layers because, conceptually, X usually spawns things that are (or become) X. The return type of a part creation operation would not be a MasterXaction (pointer) either -- it would be the type of the created part.

* A master transaction might spawn (as in "trigger" or "create") other master transactions. "Spawn" may be an acceptable name of a method in that special case, and the return type would be MasterXaction (pointer), but the implementation and (rare) use cases for such a method would be completely different from what is happening in this PR AFAICT.

>> Alex: The concept of a layer is also real and useful (in many areas), but a layer is not a master transaction.

> Amos: Agreed. "master" concept AFAIK permits the possible existence of many transactions and layers that combine to form a "master transaction".

I think we can start making good progress if you also agree that MasterXaction is a master transaction. Yes, the existing code does not use that class 100% correctly in several cases, but that is probably true for _any_ MasterXaction definition, and all those wrinkles are not particularly relevant to this disagreement AFAICT.


>> Alex: In general, I discourage folks from coding serious architectural changes affecting the entire Squid without at least some form of a preliminary agreement.

> Amos: Agreed. I thought we had consensus so far as this change goes :( 

I was very surprised to learn that you thought that MasterXaction was something other than master transaction.

 
> My planned "stage 3" future PR(s) should be moving Server state data/pointers from being ALE member to MXA member and code to use MXA directly where possible, or via ALE->xaction where not possible. Eventually (stage 4+) ALE becomes empty and logging is various way(s) to display MXA content as text.

ALE is unlikely to become empty/disappear because it (or an equivalent entity) is needed to represent, for example, two ICAP transactions that are parts of a single master transaction but get two icap_log access records/entries.

IMO, we should sketch the ALE/MasterXaction transition/evolution and reach consensus regarding primary issues before making adjustments that depend on that direction. I am not yet sure whether this PR needs this consensus.


> Alex: If we cannot agree on the architecture or even terminology, there may be another way to address my concern. If you can explain to me what this PR is meant to accomplish from Squid functionality changes, perhaps I can suggest a way to implement those changes without violating (my understanding of) the MasterXaction concept.

> Amos: Hopefully the above clears up anywhere we may not have aligned correctly so far.

The conflict has become more clear, thank you. I hope I can convince you to treat MasterXaction as a master transaction, but that remains to be seen. And, FWIW, I do not know what this PR is trying to accomplish/change (if anything) as far as Squid _functionality_ is concerned.

> * This PR is meant to enable logic in the class Server objects to know which MXA they should (in future) store their state in.

Pipelining complexities aside, there should be no choice here: For any Server (object), there is only one (current) master transaction (object).

> * Since Server objects are highly involved with layering one "master transaction" must be able to be represented as a collection of layers and/or sequence of transactions.

A master transaction is more than just a collection of parts or sequence of transactions -- it also has information shared by many of those parts, but I do not think that distinction is important here as long as we agree that MasterXaction is the class that represents that master transaction concept.

> For discussion clarity:  I am not using the term nor concept of "sub-master". I am using "sub-protocol" and sub-layer in reference to the layering property of protocol transactions _within_ a "master".

Why add "sub" at all? Why not just say "layer" or "part" of a master transaction?

[the remark this answers is not in the document]
Comment (yadij) [reply src/MasterXaction.h:66]: This discussion has gotten very side tracked from the initial "request" which just contains comments about opinion, no actual change request.

Since this is apparently stalled on terminology disagreements instead of code specifics I am going to resolve this thread as not an implementation issue and leave the discussions to another time.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/MasterXaction.h:66]: This change request is important and has not been resolved. Due to WIP PR status and many related uncertainties, it is somewhat dangerous to word this change request in terms of specific code changes, but since you insist, let's start with this:

MasterXaction::spawnChildLayer() and similar MasterXaction methods should not create (and return) a new MasterXaction object (for the reasons detailed in this change request).
